astropy/table/table.py

```diff
@@ -1795,12 +1755,58 @@ def add_column(self, col, index=None, name=None, rename_duplicate=False, copy=Tr
 
         To add several columns use add_columns.
         """
-        if index is None:
-            index = len(self.columns)
-        if name is not None:
-            name = (name,)
+        if default_name is None:
+            default_name = 'col{}'.format(len(self.columns))
+
+        # If value doesn't have a dtype and won't be added as a mixin then
+        # convert to a numpy array.
+        if not hasattr(col, 'dtype') and not self._is_mixin_for_table(col):
+            col = np.asarray(col)
+
+        # Convert col data to acceptable object for insertion into self.columns.
+        # Note that along with the lines above and below, this allows broadcasting
+        # of scalars to the correct shape for adding to table.
+        col = self._convert_data_to_col(col, name=name, copy=copy,
+                                        default_name=default_name)
+
+        # Make col data shape correct for scalars.  The second test is to allow
+        # broadcasting an N-d element to a column, e.g. t['new'] = [[1, 2]].
+        if ((len(col.shape) == 0 or (len(col.shape) > 1 and col.shape[0] == 1))
```

Comment:
Now I'm confused again, if we allow `t['new'] = [[1, 2]]` to give a `(n, 2)` shaped column, then why not let `t['new'] = [1]` give a `(n,)` shape column? It thus mean that scalar and shape `(1,)` are treated as the same, but I think that's OK.

Reply:
Overall, I now think I'd go for `if len(self) > 0 and (col.shape == () or col.shape[0] == 1):`

Reply:
OK will do.  I put the `len(self) > 0` test second because I think that will rarely fail.

Reply:
OK, makes sense.